### DISCUSSION TITLE
Mirth Channel Not Separating Segments

### DISCUSSION THREAD

**Discussion by jfldosset** [nextgenhealthcare/connect#5348] [Q&A] [answered]
I have an issue with multiple channels I have built.  With the help of the amazing folks at the Mirth Slack channel I have been able to trace the problem to the channel not separating out the segments.  The effect of this is that in the transformation stage the message appears as one large MSH segment.  So I cannot use any transforms to move data.  however it does send the message intact to its destination.  

steps for Repro
0. import channel provide below named simpleHL7toHL7
1. download sample messages
2. go to dashboard
3. right click on channel
4. click send message
5. open text file
7. select sample message you downloaded
8. process message
9. double click on channel
10. click on the source half of the message
11. go to the messages tab
12. click on the "transformed" radio button
13. observe one giant MSH segment

Originally I thought it was another issue as variables and transforms involving cross-walking data were not working.  But they couldn't work if the transformed message was one giant MSH segment and I was trying to get something out of PID.  

simple HL7 channel
[simpleHL7toHL7.txt](https://github.com/nextgenhealthcare/connect/files/9311982/simpleHL7toHL7.txt)

3 hl7 test messages from different sources that all fail to separate into segments in the transformation stage

[dduck hl7.txt](https://github.com/nextgenhealthcare/connect/files/9311997/dduck.hl7.txt)
[2ndhl7new.txt](https://github.com/nextgenhealthcare/connect/files/9311998/2ndhl7new.txt)
[3ndhl7new.txt](https://github.com/nextgenhealthcare/connect/files/9311999/3ndhl7new.txt)

an example of what the failed messages look like in the transform stage
[failed transform.txt](https://github.com/nextgenhealthcare/connect/files/9312032/failed.transform.txt)

thanks guys, I have been hitting my head on this one for about 4 days now.

**Comment by pacmano1** [comment]
Works fine on 4.0.1

**Reply by pacmano1** [reply]
Well, you have ^Ms in your file as shown by vi.   Those are new lines in vi,   you need carraige returns.

**Reply by pacmano1** [reply]
This:

![image](https://user-images.githubusercontent.com/44065187/184241884-a1ea18ce-95bc-4e7d-a96d-8449f3d6c284.png)

Should be:

![image](https://user-images.githubusercontent.com/44065187/184242234-48d425fa-a00f-45f0-83cb-fde1991de023.png)

or you can futz with the segment delimetes under the set data types button (you may have accidentally done that already)

**Reply by jfldosset** [reply]
for me in notepad++ it comes across as carriage returns \r which is the delimiter

![image](https://user-images.githubusercontent.com/68968384/184242733-9d3dc7cc-4e5a-40c3-9d62-5426e0414c70.png)



i've also tested with \r\n, \n, and \r\r\r for giggles

**Reply by jfldosset** [reply]
also using the "convert line breaks" and not using "convert line breaks"

**Reply by pacmano1** [reply]
Send this message to your channel, just cut and paste it:

```hl7
MSH|^~\&|AccMgr|1|||20050110045504||ADT^A01|599102|P|2.3|||
EVN|A01|20050110045502|||||
PID|1||10006579^^^1^MRN^1||DUCK^DONALD^D||19241010|M||1|111 DUCK ST^^FOWL^CA^999990000^^M|1|8885551212|8885551212|1|2||40007716^^^AccMgr^VN^1|123121234|||||||||||NO NK1|1|DUCK^HUEY|SO|3583 DUCK RD^^FOWL^CA^999990000|8885552222||Y||||||||||||||
PV1|1|I|PREOP^101^1^1^^^S|3|||37^DISNEY^WALT^^^^^^AccMgr^^^^CI|||01||||1|||37^DISNEY^WALT^^^^^^AccMgr^^^^CI|2|40007716^^^AccMgr^VN|4|||||||||||||||||||1||G|||20050110045253||||||
GT1|1|8291|DUCK^DONALD^D||111^DUCKST^^FOWL^CA^999990000|8885551212||19241010|M||1|123121234||||Cartoon Ducks Inc|111^DUCK ST^^FOWL^CA^999990000|8885551212||PT|
DG1|1|I9|71596^OSTEOARTHROS NOS-L/LEG ^I9|OSTEOARTHROS NOS-L/LEG ||A| IN1|1|MEDICARE|3|MEDICARE|||||||Cartoon Ducks Inc|19891001|||4|DUCK^DONALD^D|1|19241010|111^DUCK ST^^FOWL^CA^999990000|||||||||||||||||123121234A||||||PT|M|111 DUCK ST^^FOWL^CA^999990000|||||8291
IN2|1||123121234|Cartoon Ducks Inc|||123121234A|||||||||||||||||||||||||||||||||||||||||||||||||||||||||8885551212
IN1|2|NON-PRIMARY|9|MEDICAL MUTUAL CALIF.|PO BOX 94776^^HOLLYWOOD^CA^441414776||8003621279|PUBSUMB|||Cartoon Ducks Inc||||7|DUCK^DONALD^D|1|19241010|111 DUCK ST^^FOWL^CA^999990000|||||||||||||||||056269770||||||PT|M|111^DUCK ST^^FOWL^CA^999990000|||||8291
IN2|2||123121234|Cartoon Ducks Inc||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||8885551212
IN1|3|SELF PAY|1|SELF PAY|||||||||||5||1
```

**Reply by jfldosset** [reply]
unfortunately still one big MSH segment 

`  </MSH.380>
        <MSH.381>
            <MSH.381.1>SELF PAY</MSH.381.1>
        </MSH.381>
        <MSH.382/>
        <MSH.383/>
        <MSH.384/>
        <MSH.385/>
        <MSH.386/>
        <MSH.387/>
        <MSH.388/>
        <MSH.389/>
        <MSH.390/>
        <MSH.391/>
        <MSH.392>
            <MSH.392.1>5</MSH.392.1>
        </MSH.392>
        <MSH.393/>
        <MSH.394>
            <MSH.394.1>1</MSH.394.1>
        </MSH.394>
    </MSH>
</HL7Message>
`

**Reply by pacmano1** [reply]
![image](https://user-images.githubusercontent.com/44065187/184243841-32642f11-63f5-4178-9ad0-cfb258ab8fa8.png)

![image](https://user-images.githubusercontent.com/44065187/184243909-95f7472d-2ba7-4a45-a975-d12265f0b78e.png)

**Reply by mjtrac** [reply]
You should know that different OS's use different values for line endings.  As pacmano says, the HL7 spec says segments end with the carriage return, decimal 13, hex 0x0D, shown often as "\r".  I believe ^M is also a representation for carriage return.  If you are seeing one long MSH segment, the most likely explanation is that there's a mismatch between the channel's end-of-segment setting and the values coming into your channel.  If you're on a UNIXy system, including OSX, you can use octal dump (type "man od" in a terminal for usage info) to examine your file as hex values.  I'm sure Windows has something similar.  To avoid future confusion, you probably want to continue to follow HL7 rules in your channel.  But if the channel is set up already to expect an alternate line ending, you can use the Preprocessor script to change instances of whatever you now have coming in to whatever is set in the Set Data Types dialog.  Hope this helps; there's also an extensive Wikipedia entry for "newline."

![image](https://user-images.githubusercontent.com/2815700/184243648-405612b5-30d6-4191-b894-7c338cc4522d.png)

**Reply by jfldosset** [reply]
yeap.  that is exactly what it SHOULD look like but I get this instead.
![image](https://user-images.githubusercontent.com/68968384/184245094-39cf0986-dee7-4963-8180-ba34c6bb4c25.png)


i find it gratifying and frustrating my channel is working for others

**Reply by jfldosset** [reply]
@mjtrac   you might be on to something.   I am on a windows machine and my mirth server is on an AWS EC2 instance running some *nix variant.   I have tried a \n only, but maybe that *nix is using something weird.....i dunno.  

a great lead thanks!

**Reply by pacmano1** [reply]
The fact the MSH "merges" means mirth is not seeing your delimeters for the end of the segment, it's that simple.   Futz with the line endings to you sort it out.   If I were you, ssh to the AWS instance,  edit a file (paste the file above I posted), then have mirth read the file from the file system.

**Reply by siddharthsharma9258** [reply]
> unfortunately still one big MSH segment
> 
> ` </MSH.380> <MSH.381> <MSH.381.1>SELF PAY</MSH.381.1> </MSH.381> <MSH.382/> <MSH.383/> <MSH.384/> <MSH.385/> <MSH.386/> <MSH.387/> <MSH.388/> <MSH.389/> <MSH.390/> <MSH.391/> <MSH.392> <MSH.392.1>5</MSH.392.1> </MSH.392> <MSH.393/> <MSH.394> <MSH.394.1>1</MSH.394.1> </MSH.394> </MSH> </HL7Message>`

Ofcourse, cos the message cannot be broken down into segments and such. It is looking especially for new line character. There are only 2 options - \r\n or \n (no \r\r\r for giggles :))

**Reply by jfldosset** [reply]
so after some testing apparently whatever madman created our *nix install we are using decide that $ should be the EOL character of choice and not....anything standard and is shoving that in there if I use anything halfway normal as an EOL character such as \r or \n or any combination of those.  I even tried \n\n\n and replaced them with $$$.

so that will have to be dealt with


![image](https://user-images.githubusercontent.com/68968384/184432593-b5844792-afc3-465a-b9f3-2a2e7739acbe.png)

**Reply by pacmano1** [reply]
No.

In vi `$` is the end of line as displayed by `:set list `

**Reply by ChristopherSchultz** [reply]
^M is CR, not LF. The message looks fine.

**Comment by jfldosset** [comment]
ok so i did some testing 

i replace the carriage returns with %%% as my carriage return in the channel and in the message.

now it works.

my current theory is that when i right click on the channel and manually load the file it must be substituting the windows carriage return for \n.  I'll try loading it from an s3 bucket since i have that ready to go on a channel and then from a file on the AWS instance.

will update tomorrow.  time to go home

**Comment by jfldosset** [comment, marked as answer]
I have a resolution to this.    

Somewhere in the linux that our Mirth is running on is an issue with carriage returns and making control-M (^M) into possibly a literal ^ and M.

`var newmessage = message.replace(/(\\n\^M)/, "");
var newmessage = message.replace(/(\\r\^M)/, "");
var newmessage = message.replace(/(\\r\n\^M)/, "");
var newmessage = message.replace(/(\\n\r\^M)/, "");
return newmessage;`

depending on How I sent across the messages one of the above 4 errors would occur.  yes I will be streamlining the regex but this was for testing purposes.  once these were in the preprocessor I had no further troubles with hl7 messages from any source.

thank you everyone for all your help

**Reply by jfldosset** [reply]
yeah so oddly enough somewhere in the interaction of the linux system and mirth it is adding one of the 4 options Extra on top of the regular \n  that should be there depending on how the message is getting to the channel: file from the linux server, s3 file, copy paste, etc.

 thanks for reminding me about the g i need to test that further

**Reply by tonygermano** [reply]
There should not be a `\n` there at all. The hl7 segment delimiter is `\r`. You should never change this in the channel settings unless you know that you are getting really bad data and it's the only way to make it work. If you are getting `\n` or `\r\n` instead of `\r` you still should not change the segment delimiter, and instead just make sure the Convert Line Breaks option is enabled.

**Reply by tonygermano** [reply]
And I highly (highly!) doubt that your linux distribution has anything to do with converting control characters to string literals in your messages.

**Reply by jfldosset** [reply]
Convert line breaks wasn't working until I inserted those lines in the preprocessor and we are definitely getting an \n in addition to the \r on any message that that passes through that linux instance.  I even installed a Mirth server and tested it on my windows laptop, I had zero issues with any of the channels or messages.   So if its not the messages and its not mirth and its not the channels then it is the linux instance and if it is not linux then it is something something AWS.   I'm open to any further testing ideas but for now I will leave this as a workaround as I spent almost 2 weeks on it.
